source/Weather.cpp

```diff
@@ -70,29 +71,50 @@ const Point &Weather::Origin() const
 
 
 // Create any environmental effects and decrease the lifetime of this weather.
-void Weather::Step(vector<Visual> &visuals)
+void Weather::Step(vector<Visual> &visuals, const Point &center)
 {
 	// Environmental effects are created by choosing a random angle and distance from
 	// their origin, then creating the effect there.
 	double minRange = hazard->MinRange();
 	double maxRange = hazard->MaxRange();
+	double effectMultiplier = currentStrength;
 
-	// Estimate the number of visuals to be generated this frame.
-	// MAYBE: create only a subset of possible effects per frame.
-	int totalAmount = 0;
-	for(auto &&effect : hazard->EnvironmentalEffects())
-		totalAmount += effect.second;
-	totalAmount *= currentStrength;
-	visuals.reserve(visuals.size() + totalAmount);
-
-	for(auto &&effect : hazard->EnvironmentalEffects())
-		for(int i = 0; i < effect.second * currentStrength; ++i)
-		{
-			Point angle = Angle::Random().Unit();
-			double magnitude = (maxRange - minRange) * sqrt(Random::Real());
-			Point pos = origin + (minRange + magnitude) * angle;
-			visuals.emplace_back(*effect.first, std::move(pos), Point(), Angle::Random());
-		}
+	// If a hazard is system-wide, the max range becomes the edge of the screen,
+	// and the number of effects drawn is scaled accordingly.
+	if(hazard->SystemWide())
```

Comment:
Maybe add `&& maxRange > 0.` ?

Reply:
Sure, although for doubles it doesn't matter to divide by 0 you get infinite.
However, in this specific case that could mean an infinite loop later on so I added it.

Also may as well spare the calculations if we have a negative range and it won't draw anything anyway, as you suggested `> 0`, instead of just checking if its different than 0.